src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,685 @@
+
+        foreach ((array) $needle as $n) {
+            $n = (string) $n;
+            $j = $this->indexOf($n, $offset);
+
+            if (null !== $j && $j < $i) {
+                $i = $j;
+                $str->string = $n;
+            }
+        }
+
+        if (\PHP_INT_MAX === $i) {
+            return $str;
+        }
+
+        if ($includeNeedle) {
+            $i += $str->length();
+        }
+
+        return $this->slice(0, $i);
+    }
+
+    /**
+     * @param string|string[] $needle
+     *
+     * @return static
+     */
+    public function beforeLast($needle, bool $includeNeedle = false, int $offset = 0): self
+    {
+        $str = clone $this;
+        $str->string = '';
+        $i = null;
+
+        foreach ((array) $needle as $n) {
+            $n = (string) $n;
+            $j = $this->indexOfLast($n, $offset);
+
+            if (null !== $j && $j > $i) {
+                $i = $offset = $j;
+                $str->string = $n;
+            }
+        }
+
+        if (null === $i) {
+            return $str;
+        }
+
+        if ($includeNeedle) {
+            $i += $str->length();
+        }
+
+        return $this->slice(0, $i);
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function camel(): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
+    {
+        $str = clone $this;
+        $str->string = trim(preg_replace('/(?:\s{2,}+|[^\S ])/', ' ', $str->string));
+
+        return $str;
+    }
+
+    /**
+     * @param string|string[] $suffix
+     */
+    public function endsWith($suffix): bool
+    {
+        if (!\is_array($suffix) && !$suffix instanceof \Traversable) {
+            throw new \TypeError(sprintf('Method "%s()" must be overridden by class "%s" to deal with non-iterable values.', __FUNCTION__, \get_class($this)));
+        }
+
+        foreach ($suffix as $s) {
+            if ($this->endsWith((string) $s)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @return static
+     */
+    public function ensureEnd(string $suffix): self
+    {
+        if (!$this->endsWith($suffix)) {
+            return $this->append($suffix);
+        }
+
+        $suffix = preg_quote($suffix);
+        $regex = '{('.$suffix.')(?:'.$suffix.')++$}D';
+
+        return $this->replaceMatches($regex.($this->ignoreCase ? 'i' : ''), '$1');
+    }
+
+    /**
+     * @return static
+     */
+    public function ensureStart(string $prefix): self
+    {
+        $prefix = new static($prefix);
+
+        if (!$this->startsWith($prefix)) {
+            return $this->prepend($prefix);
+        }
+
+        $str = clone $this;
+        $i = $prefixLen = $prefix->length();
+
+        while ($this->indexOf($prefix, $i) === $i) {
+            $str = $str->slice($prefixLen);
+            $i += $prefixLen;
```

Comment:
This is wrong, isn't it? We are already shortening the string before, so we wanna check again at the same position.

Reply:
it's correct, this is tested actually